utils/s2n_ensure.h

```diff
@@ -67,4 +67,21 @@
     }                                                                            \
   } while(0)
 
+/**
+ * `restrict` is a part of the c99 standard and will work with any C compiler. If you're trying to
+ * compile with a C++ compiler `restrict` is invalid. However GCC and Clang support the use of
+ * `restrict` when compiling C++ code using `__restrict__`. Therefore if the compiler is GCC or
+ * Clang use `__restrict__` which will behave the same as `restrict` and allow g++ and clang++ to
+ * compile it.
+ *
+ * This is helpful for the benchmarks in tests/benchmark which use Google's Benchmark library and
+ * are all written in C++.
+ *
+ * https://gcc.gnu.org/onlinedocs/gcc/Restricted-Pointers.html
+ *
+ */
+#if defined(__GNUC__) || defined(__clang__)
```

Comment:
Can you add a feature probe for this instead? See https://github.com/awslabs/s2n/pull/2559 for an example